File: mobile_v2/lib/features/messaging/data/models/message_model.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import '../../../../features/auth/data/models/user_model.dart';

part 'message_model.freezed.dart';
part 'message_model.g.dart';

@freezed
class MessageModel with _$MessageModel {
  const factory MessageModel({
    required dynamic id,
    @JsonKey(name: 'sender_id') required dynamic senderId,
    @JsonKey(name: 'receiver_id') required dynamic receiverId,
    @JsonKey(name: 'listing_id') dynamic listingId,
    required String content,
    @JsonKey(name: 'read_at') String? readAt,
    @JsonKey(name: 'created_at') String? createdAt,
    UserModel? sender,
    UserModel? receiver,
  }) = _MessageModel;

  factory MessageModel.fromJson(Map<String, dynamic> json) =>
      _$MessageModelFromJson(json);
}

@freezed
class ConversationModel with _$ConversationModel {
  const factory ConversationModel({
    required UserModel otherUser,
    required MessageModel lastMessage,
    required int unreadCount,
  }) = _ConversationModel;

  factory ConversationModel.fromJson(Map<String, dynamic> json) =>
      _$ConversationModelFromJson(json);
}

File: backend-ci4/app/Models/MessageModel.php
<?php

namespace App\Models;

use CodeIgniter\Model;

class MessageModel extends Model
{
    protected $table = 'messages';
    protected $primaryKey = 'id';
    protected $allowedFields = ['sender_id', 'receiver_id', 'listing_id', 'content', 'read_at'];
    protected $useTimestamps = true;
    protected $createdField  = 'created_at';
    protected $updatedField  = 'updated_at';
}
